modules/jwplayerTargeting.js

```diff
@@ -0,0 +1,167 @@
+import { getGlobal } from '../src/prebidGlobal.js';
+import { module } from '../src/hook.js';
+
+const segCache = {};
+let requestCount = 0;
+let requestTimeout;
+let resumeBidRequest;
+
+/*
+Prebid auctions timeout at 200ms.
+ */
+let feedFetchTimeout = 150;
+
+config.getConfig('jwTargeting', config => {
+  const targeting = config.jwTargeting;
+  // fetch media ids
+  fetchTargetingInformation(targeting);
+
+  const prefetchTimeout = targeting.prefetchTimeout;
+  if (prefetchTimeout) {
+    // prefetch timeout supersedes our default and our adjustment for bidderTimeout.
+    feedFetchTimeout = prefetchTimeout;
+    return;
+  }
+
+  const timeout = config.bidderTimeout; 
+  if (timeout < 200) {
+    // 3/4 is the ratio between 150 and 200, where 150 is our default and 200 is prebid's default auction timeout.
+    // Note auction will close at 200ms even if bidderTimeout is greater.
+    feedFetchTimeout = timeout * 3 / 4;
+  }
+});
+
+getGlobal().requestBids.before(ensureFeedRequestCompletion);
+
+export function fetchTargetingInformation(jwTargeting) {
+  const mediaIDs = jwTargeting.mediaIDs;
+  requestCount = mediaIDs.length;
+  mediaIDs.forEach(mediaID => {
+    fetchTargetingForMediaId(mediaID);
+  });
+}
+
+export function ensureFeedRequestCompletion(requestBids, bidRequestConfig) {
+  if (requestCount <= 0) {
+    requestBids.apply(this, [bidRequestConfig]);
+    return;
+  }
+  resumeBidRequest = requestBids.bind(this, bidRequestConfig);
+  requestTimeout = setTimeout(function() {
+    resumeBidRequest();
+    resumeBidRequest = null;
+    requestTimeout = null;
+  }, feedFetchTimeout);
+}
+
+/**
+ * @param bidRequest {object} - the bid which is passed to a prebid adapter for use in `buildRequests`
+ * @returns {Array<string>} - an array of jwpseg targeting segments found for the given bidRequest information
+ */
+export function getTargetingForBid(bidRequest) {
+  const jwTargeting = bidRequest.jwTargeting;
+  if (!jwTargeting) {
+    return [];
+  }
+
+  const playerID = jwTargeting.playerID;
+  let mediaID = jwTargeting.mediaID;
+  let segments = segCache[mediaID];
+  if (segments) {
+    return segments;
+  }
+
+  const player = getPlayer(playerID);
+  if (!player) {
+    return [];
+  }
+
+  const item = mediaID ? player.getPlaylist().find(item => item.mediaid === mediaID) : player.getPlaylistItem();
+  if (!item) {
+    return [];
+  }
+
+  mediaID = mediaID || item.mediaid;
+  segments = item.jwpseg;
+  if (segments && mediaID) {
+    segCache[mediaID] = segments;
+  }
+
+  return segments || [];
+}
+
+function getPlayer(playerID) {
+  const jwplayer = window.jwplayer;
+  if (!jwplayer) {
+    logError('jwplayer.js was not found on page');
+    return;
+  }
+
+  const player = jwplayer(playerID);
+  if (!player || !player.getPlaylist) {
+    logError('player ID did not match any players');
+    return;
+  }
+  return player;
+}
+
+export function fetchTargetingForMediaId(mediaId) {
+  const ajax = ajaxBuilder(1500);
```

Comment:
Should this be `feedFetchTimeout`?

Reply:
no, feedFetchTimeout is used only to pause a bid if there is a bid request being sent while the feeds are being fetched. If no bid is being sent, then the feed fetch can take longer. 
Also, even though the feed fetch may not complete by the bid timeout, we would resume the bid but we wouldn't want to cancel the feed request; we would want it to  continue because Another bid might go out in the future. 
Great question. I think i should rename `feedFetchTimeout`.